runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,292 @@
+local keymap = {}
+
+-- Have to use a global to handle re-requiring this file and losing all of the keymap.
+--  In the future, the C code will handle this.
+__VimMapStore = __VimMapStore or {}
+keymap._store = __VimMapStore
+
+keymap._create = function(f)
+  table.insert(keymap._store, f)
+  return #keymap._store
+end
+
+keymap._execute = function(id)
+  keymap._store[id]()
+end
+
+local make_mapper = function(mode, defaults)
+  return function(opts)
+    local args, map_args = {}, {}
+    for k, v in pairs(opts) do
+      if type(k) == 'number' then
+        args[k] = v
+      else
+        map_args[k] = v
+      end
+    end
+
+    local lhs = opts.lhs or args[1]
+    local rhs = opts.rhs or args[2]
+
+    local mapping
+    if type(rhs) == 'string' then
+      mapping = rhs
+    elseif type(rhs) == 'function' then
+      local func_id = keymap._create(rhs)
+
+      mapping = string.format(
+        [[:lua vim.keymap._execute(%s)<CR>]], func_id
+      )
+    end
+
+    local map_opts = vim.tbl_extend("force", defaults, map_args)
+
+    if not map_opts.buffer then
+      vim.api.nvim_set_keymap(mode, lhs, mapping, map_opts)
+    else
+      -- Clear the buffer after saving it
+      local buffer = map_opts.buffer
+      map_opts.buffer = nil
+
+      vim.api.nvim_buf_set_keymap(buffer, mode, lhs, mapping, map_opts)
+    end
+  end
+end
+
+-- Use this code to generate all the things
+function keymap._generate_keymap_codes()
```

Comment:
lua code generating lua is kinda sus, is it needed only because documentation?

Reply:
Yeah :/ Only for documentation. Otherwise I would just dynamically set the keys with all the combos. 

I can just type them all out if it makes you feel better :)

Reply:
Alternatively enhance doc generation to only do one step of generation and not two :)